src/box/box.cc

```diff
@@ -4296,3 +4297,264 @@ box_read_ffi_enable(void)
+			  uint32_t name_len, uint32_t fieldno)
+{
+	return constraint_drop(space_id, name, name_len, fieldno, CONSTR_FUNC);
+}
+
+int
+box_tuple_foreign_key_drop(uint32_t space_id, const char *name,
+			   uint32_t name_len)
+{
+	return constraint_drop(space_id, name, name_len, UINT32_MAX,
+			       CONSTR_FKEY);
+}
+
+int
+box_field_foreign_key_drop(uint32_t space_id, const char *name,
+			   uint32_t name_len, uint32_t fieldno)
+{
+	return constraint_drop(space_id, name, name_len, fieldno, CONSTR_FKEY);
+}
+
+/* Value of new constraint or foreign key. */
+union constraint_value {
+	/* ID of a function for constraint. */
+	uint32_t func_id;
+	struct {
+		/* Mapping for complex foreign key. */
+		const char *mapping;
+		/* Foreign space ID. */
+		uint32_t space_id;
+		/* Fieldno of foreign field for simple foreign key. */
+		uint32_t fieldno;
+	};
+};
+
+/**
+ * Create constraint or foreign key. If fieldno is equal to UINT32_MAX then the
+ * tuple constraint or tuple foreign key is created, otherwise the field
+ * constraint or field foreign key is created.
+ */
+static int
+constraint_create(uint32_t space_id, const char *name, uint32_t name_len,
+		  uint32_t fieldno, const union constraint_value *value,
+		  enum tuple_constraint_type type)
+{
+	const struct space *space = space_by_id(space_id);
+	uint32_t total_count;
+	struct tuple_constraint_def *cdefs;
+	if (fieldno != UINT32_MAX) {
+		assert(fieldno < space->def->field_count);
+		total_count = space->def->fields[fieldno].constraint_count;
+		cdefs = space->def->fields[fieldno].constraint_def;
+	} else {
+		total_count = space->def->opts.constraint_count;
+		cdefs = space->def->opts.constraint_def;
+	}
+	bool is_empty = true;
+	for (uint32_t i = 0; i < total_count && is_empty; ++i)
+		is_empty = cdefs[i].type != type;
+
+	char key[16];
+	char *key_end = key + mp_format(key, 16, "[%u]", space_id);
+	const char *type_str = tuple_constraint_type_strs[type];
+	const char *path = fieldno != UINT32_MAX ?
+			   tt_sprintf("format[%d].%s", fieldno + 1, type_str) :
+			   tt_sprintf("flags.%s", type_str);
+	/*
+	 * Even if there were no constraints of this type, it is possible that
+	 * _space contains a non-empty field of this type with an empty map as
+	 * its value, which affects the update operation.
+	 */
+	if (is_empty) {
+		struct tuple *tuple;
+		if (box_index_get(BOX_SPACE_ID, 0, key, key_end, &tuple) != 0)
+			return -1;
+		uint32_t path_len = strlen(path);
+		uint32_t path_hash = field_name_hash(path, path_len);
+		const char *field =
+			tuple_field_raw_by_full_path(tuple_format(tuple),
+						     tuple_data(tuple),
+						     tuple_field_map(tuple),
+						     path, path_len, path_hash,
+						     TUPLE_INDEX_BASE);
+		is_empty = field == NULL;
+	}
+
+	struct region *region = &fiber()->gc;
+	uint32_t used = region_used(region);
+	const char *val;
+	if (type == CONSTR_FUNC) {
+		val = mp_format_on_region(region, "%u", value->func_id);
+	} else if (fieldno != UINT32_MAX) {
```

Comment:
I suggest to assert here that the type is either FUNC or FKEY.

Reply:
Thanks, added.